examples/kokkos/view/example_restrict.cpp

```diff
@@ -0,0 +1,243 @@
+#include "Kokkos_Core.hpp"
+#include "Kokkos_Profiling_ProfileSection.hpp"
+#include "Kokkos_Profiling_ScopedRegion.hpp"
+
+namespace reprospect::examples::kokkos::view
+{
+#define OPERATION(dst, src_a, src_b, index)    \
+    dst[index]  = src_a[index] + src_b[index]; \
+    dst[index] += src_a[index] + src_b[index];
+
+template <typename T>
+__global__ void global_kernel(
+    T* __restrict__ const dst,
+    const T* __restrict__ const src_a, const T* __restrict__ const src_b,
+    const unsigned int size
+) {
+    const auto index = blockIdx.x * blockDim.x + threadIdx.x;
+    if (index < size) {
+        OPERATION(dst, src_a, src_b, index)
+    }
+}
+
+template <typename T>
+struct FunctorRestrictMember
+{
+    const T* __restrict__ src_a;
+    const T* __restrict__ src_b;
+          T* __restrict__ dst;
+
+    template <std::integral U>
+    KOKKOS_FUNCTION void operator()(const U index) const {
+        OPERATION(dst, src_a, src_b, index)
+    }
+};
+
+template <typename T>
+struct FunctorRestrictRecast
+{
+    const T* src_a;
+    const T* src_b;
+          T* dst;
+
+    template <std::integral U>
+    KOKKOS_FUNCTION void operator()(const U index) const {
+        [index](T* __restrict__ const d, const T* __restrict__ const a, const T* __restrict__ const b) {
+            OPERATION(d, a, b, index)
+        }(dst, src_a, src_b);
```

Comment:
It seems that:

```
    template <std::integral U>
    KOKKOS_FUNCTION void operator()(const U index) const {
        T* __restrict__ const dst = this->dst;
        const T* __restrict__ const src_a = this->src_a;
        const T* __restrict__ const src_b = this->src_b;
        OPERATION(dst, src_a, src_b, index)
    }
```
works as well. 

Some speculation: perhaps there is a way to push this to making it work with view. We'd have to copy construct the views into restricted views inside the call operator. The restricted views would have to store their pointer with the restrict keyword. The accessor wouldn't have to be annotated. Checking how a restricted view currently stores its data pointer would be a first step.

Reply:
The speculation doesn't work. It can't work through an accessor.